source_gen/lib/src/builder.dart

```diff
@@ -144,24 +152,51 @@ class _Builder extends Builder {
   }
 }
 
+/// A [Builder] which generates `part of` files.
+///
+/// Generated files will be prefixed with a `partId` to ensure multiple
+/// [SharedPartBuilder]s can produce non conflicting `part of` files.
+class SharedPartBuilder extends _Builder {
+  /// Wrap [generators] as a [Builder] that generates `part of` files.
+  ///
+  /// [partId] indicates what files will be created for each `.dart`
+  /// input. This extension should be unique as to not conflict with other
+  /// [SharedPartBuilder]s. The resulting file will be of the form
+  /// `<generatedExtension>.g.part`. If any generator in [generators] will
+  /// create additional outputs through the [BuildStep] they should be indicated
+  /// in [additionalOutputExtensions].
+  ///
+  /// [formatOutput] is called to format the generated code. Defaults to
+  /// [DartFormatter.format].
+  SharedPartBuilder(List<Generator> generators, String partId,
+      {String formatOutput(String code),
+      List<String> additionalOutputExtensions = const []})
+      : super(generators,
+            formatOutput: formatOutput,
+            generatedExtension: '$partId.g.part',
```

Comment:
should have validation here for `partId` -> `generatedExtension`.

User is required to provide a `.` before – that's broken. I'd argue that it should be `A-Za-z0-9_\.`with no before/after `.`